main.cu

```diff
@@ -174,144 +133,187 @@ __device__ inline int getTreeHeight(int x, int z) {
 
-    if(res) {
-        int index = atomicAdd(num_starts, 1);
-        starts[index] = tree_start;
+    if (res) {
+        int index = atomicAdd(num_tree_starts, 1);
+        tree_starts[index] = tree_start;
     }
 }
 
-__global__ void doWork(int* num_starts, Random* tree_starts, int* num_seeds, ulong* seeds, int gpu_search_back_count) {
-    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < *num_starts; i += blockDim.x * gridDim.x) {
+__global__ void doCalcRandStarts(int* num_tree_starts,
+                                 Random* tree_starts,
+                                 int* num_rand_starts,
+                                 Random* rand_starts,
+                                 int gpu_search_back_count) {
+    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < *num_tree_starts; i += blockDim.x * gridDim.x) {
         Random tree_start = tree_starts[i];
 
         for (int treeBackCalls = 0; treeBackCalls <= gpu_search_back_count; treeBackCalls++) {
-            Random start = (tree_start * search_back_multipliers[treeBackCalls] + search_back_addends[treeBackCalls]) & RANDOM_MASK;
-            Random rand = start;
-
-            bool this_res = true;
-
-            if(random_next_int(&rand, 10) == 0)
-                continue;
-
-            char generated_tree[16][2];
-            memset(generated_tree, 0x00, sizeof(generated_tree));
+            ulong multiplier = search_back_multipliers[treeBackCalls];
+            ulong addend = search_back_addends[treeBackCalls];
+            Random start = (tree_start * multiplier + addend) & RANDOM_MASK;
 
-            int treesMatched = 0;
-            bool any_population_matches = false;
-            for (int treeAttempt = 0; treeAttempt <= MAX_TREE_ATTEMPTS; treeAttempt++) {
-                int treeX = random_next(&rand, 4);
-                int treeZ = random_next(&rand, 4);
-                int wantedTreeHeight = getTreeHeight(treeX, treeZ);
-                int treeHeight = random_next_int(&rand, 3) + 4;
-
-                char& boolpack = generated_tree[treeX][treeZ / 2];
-                const char mask = 1 << (treeZ % 8);
+            Random rand = start;
+            if (random_next_int(&rand, 10) != 0) {
+                int index = atomicAdd(num_rand_starts, 1);
+                rand_starts[index] = start;
+            }
+        }
+    }
+}
 
-                if (treeHeight == wantedTreeHeight && !(boolpack & mask)) {
-                    treesMatched++;
-                    boolpack |= mask;
-                    advance_16(rand);
-                }
+__global__ void doCalcChunkSeed(int* num_rand_starts, 
+                                Random* rand_starts,
+                                int* num_seeds,
+                                ulong* seeds) {
+    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < *num_rand_starts; i += blockDim.x * gridDim.x) {
+        Random rand = rand_starts[i];
+        Random start = rand;
+
+        // Advance 10
```

Comment:
This line doesn't advance by 10, it advances by 1 and returns a value from 0-9. This line can be omitted by changing the previous step to put the values of `rand` in the array, rather than `start`. Then, when adding to the seed list, advance by -3760 rather than -3759.

Reply:
Hmmm, I did think of this however of was unsure about the mechanics of advance and if you could only advance by that constant.

That being said how are the values for advance calculated, I would need the new values for m3760.